core/query/mixingQuery.ml

```diff
@@ -0,0 +1,1551 @@
+open Utility
+open CommonTypes
+open Var
+open Errors
+
+exception DbEvaluationError of string
+
+let internal_error message =
+  Errors.internal_error ~filename:"query/query.ml" ~message
+
+let runtime_type_error error =
+  internal_error
+    ("Runtime type error: " ^ error ^ ".\n" ^
+     "This should not happen if the type system / checker is correct. " ^
+     "Please file a bug report.")
+
+let query_error fmt =
+  let error msg = raise (DbEvaluationError msg) in
+    Printf.kprintf error fmt
+
+module Lang =
+struct
+
+  type base_type = | Bool | Char | Float | Int | String
+
+  type tag = int
+      [@@deriving show]
+
+  type t =
+      | For       of tag option * (Var.var * t) list * t list * t
+      | If        of t * t * t
+      | Table     of Value.table
+      | Database  of (Value.database * string)
+      | Singleton of t
+      | Concat    of t list
+      | Dedup     of t
+      | Prom      of t
+      | Record    of t StringMap.t
+      | Project   of t * string
+      | Erase     of t * StringSet.t
+      | Variant   of string * t
+      | XML       of Value.xmlitem
+      | Apply     of t * t list
+      | Closure   of (Ir.var list * Ir.computation) * env
+      | Case      of t * (binder * t) StringMap.t * (binder * t) option
+      | Primitive of string
+      (* | Var       of Var.var * Types.datatype StringMap.t *)
+      | Var       of Var.var * Types.datatype
```

Comment:
Could this be unified with Query.t by abstracting over the variable annotation type?  (Make a note of this as a comment in the code for future reference)

Reply:
Unifying the two query types by changing the `Var` annotation to `Types.datatype` everywhere seems to have done the trick. It took a lot of work, but all the alternatives were less convincing...